crates/bevy_tasks/src/task_pool.rs

```diff
@@ -310,23 +346,61 @@ impl TaskPool {
                     results
                 };
 
-                // Pin the futures on the stack.
-                pin!(get_results);
+                let tick_task_pool_executor = tick_task_pool_executor || self.threads.is_empty();
+                if let Some(thread_ticker) = thread_executor.ticker() {
```

Comment:
I'm growing increasingly concerned with the complexity of this function. It's getting to the point where it's legit hard to follow the logic within (on par with `Query::par_for_each`). Can we chunk this up in any way?

Reply:
not in a very clean way. The problem is all the executor futures are actually different types. I could potentially make 3 different functions for getting the executors. Would at least look a little cleaner here, but I'm not sure if it's actually better. Let me know what you think:

```rust
            future::block_on(async move {
                let get_results = async {
                    let mut results = Vec::with_capacity(spawned_ref.len());
                    while let Ok(task) = spawned_ref.pop() {
                        results.push(task.await.unwrap());
                    }

                    results
                };

                let tick_task_pool_executor = tick_task_pool_executor || self.threads.is_empty();
                if let Some(thread_ticker) = thread_executor.ticker() {
                    if tick_task_pool_executor {
                        Self::execute_local_global(thread_ticker, executor, get_results).await
                    } else {
                        Self::execute_local(thread_ticker, get_results).await
                    }
                } else if tick_task_pool_executor {
                    Self::execute_global(executor, get_results).await
                } else {
                    get_results.await
                }
            })
        }
    }

    async fn execute_local_global<'scope, 'ticker, T>(
        thread_ticker: ThreadExecutorTicker<'scope, 'ticker>,
        executor: &'scope async_executor::Executor<'scope>,
        get_results: impl Future<Output = Vec<T>>,
    ) -> Vec<T> {
        // we restart the executors if a task errors. if a scoped
        // task errors it will panic the scope on the call to get_results
        let execute_forever = async move {
            loop {
                let tick_forever = async {
                    loop {
                        thread_ticker.tick().await;
                    }
                };

                // we don't care if it errors. If a scoped task errors it will propagate
                // to get_results
                let _result = AssertUnwindSafe(executor.run(tick_forever))
                    .catch_unwind()
                    .await
                    .is_ok();
            }
        };

        execute_forever.or(get_results).await
    }

    async fn execute_local<'scope, 'ticker, T>(
        thread_ticker: ThreadExecutorTicker<'scope, 'ticker>,
        get_results: impl Future<Output = Vec<T>>,
    ) -> Vec<T> {
        let execute_forever = async {
            loop {
                let tick_forever = async {
                    loop {
                        thread_ticker.tick().await;
                    }
                };

                let _result =
                    AssertUnwindSafe(tick_forever).catch_unwind().await.is_ok();
            }
        };

        execute_forever.or(get_results).await
    }

    async fn execute_global<'scope, T>(
        executor: &'scope async_executor::Executor<'scope>,
        get_results: impl Future<Output = Vec<T>>,
    ) -> Vec<T> {
        let execute_forever = async {
            loop {
                let _result =
                    AssertUnwindSafe(executor.run(std::future::pending::<()>()))
                        .catch_unwind()
                        .await
                        .is_ok();
            }
        };

        execute_forever.or(get_results).await
    }
```

Reply:
This generally looks better to me, since the distance between the branch is smaller now and the individual approaches to executing it are isolated.

It does make one raise a few questions about the spooky action at a distance a caught unwind would cause though due to the AssertUnwindSafe usage, but as mentioned before that probably isn't too much of an issue.